hc5/src/main/java/feign/hc5/AsyncApacheHttp5Client.java

```diff
@@ -0,0 +1,183 @@
+/**
+ * Copyright 2012-2020 The Feign Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package feign.hc5;
+
+import org.apache.hc.client5.http.async.methods.SimpleHttpRequest;
+import org.apache.hc.client5.http.async.methods.SimpleHttpResponse;
+import org.apache.hc.client5.http.config.Configurable;
+import org.apache.hc.client5.http.config.RequestConfig;
+import org.apache.hc.client5.http.impl.async.CloseableHttpAsyncClient;
+import org.apache.hc.client5.http.impl.async.HttpAsyncClients;
+import org.apache.hc.client5.http.protocol.HttpClientContext;
+import org.apache.hc.core5.concurrent.FutureCallback;
+import org.apache.hc.core5.http.ContentType;
+import org.apache.hc.core5.http.Header;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import feign.*;
+import feign.Request.Options;
+
+/**
+ * This module directs Feign's http requests to Apache's
+ * <a href="https://hc.apache.org/httpcomponents-client-5.0.x/index.html">HttpClient 5</a>. Ex.
+ *
+ * <pre>
+ * GitHub github = Feign.builder().client(new ApacheHttp5Client()).target(GitHub.class,
+ * "https://api.github.com");
+ */
+/*
+ */
+public final class AsyncApacheHttp5Client<C> implements AsyncClient<C> {
```

Comment:
I think here this should actually be:
```suggestion
public final class AsyncApacheHttp5Client implements AsyncClient<HttpClientContext> {
```
This then enables the user to create their own external context as part of AsyncFeign - thus, they can provide their own CookieStore for example and have session management.

Reply:
I'm not convinced we need to add another construct like the `Context` when we have `Request` already, which represents the canonical information required for a `Client` to execute a request.  It's already thread-safe and immutable.  We should consider extending `Request` before adding another concept.

Another counter point, this `Context` provides users a way to circumvent our internal separation and, if history is any indication, will become a cause for confusion.  If you decide to continue to use this `Context`, I strongly suggest that we make it part of our private API usage.

Cookies and Session management can be something that we should support first at the `Contract` level, ensuring compatibility throughout the ecosystem and preserve the domain separation Feign is good at, which we should address separately from this PR.

Reply:
For now I will keep the Context, but I agree that is something we need to revisit

Reply:
agreed.